File: 2.3-2 merge-sort__no-sentinel/code.py
def sort(A):
    mergeSort(A, 0, len(A))

def mergeSort(A, p, r):
    q = ( p + r ) // 2
    if p < q:
        mergeSort(A, p, q)
        mergeSort(A, q, r)
        merge(A, p, q, r)

def merge(A, p, q, r):
    L = []
    R = []
    for i in range(p, q):
        L.append(A[i])
    for i in range(q, r):
        R.append(A[i])
    iL = 0
    iR = 0
    for i in range(p, r):
        if iL >= len(L):
            for j in range(iR, len(R)):
                A[i + j - iR] = R[j]
            break
        if iR >= len(R):
            for j in range(iL, len(L)):
                A[i + j - iL] = L[j]
            break
        if L[iL] > R[iR]:
            A[i] = R[iR]
            iR += 1
        else:
            A[i] = L[iL]
            iL += 1


A = [4, 3, 7, 10, 4, 34, 11, 2, 45, 62, 93, 25]
sort(A)
print(A)

File: 2.3-2 merge-sort__no-sentinel/code.js
/**
 * 
 * @param {number[]} A 
 */
function sort(A){
    mergeSort(A, 0, A.length);
}

/**
 * 
 * @param {number[]} A 
 * @param {number} p 
 * @param {number} r 
 */
function mergeSort(A, p, r){
    let q = Math.floor((p + r) / 2);
    if(q > p){
        mergeSort(A, p, q);
        mergeSort(A, q, r);
        merge(A, p, q, r);
    }
}


/**
 * 
 * @param {number[]} A 
 * @param {number} p 
 * @param {number} q
 * @param {number} r 
 */
function merge(A, p, q, r){
    let L = [];
    let R = [];
    for (let i = p; i < q ; i++) {
        L.push(A[i]);
    }
    for (let i = q; i < r ; i++) {
        R.push(A[i]);
    }
    let iL = iR = 0;
    for(let i = p; i < r; i ++){
        if (iL >= L.length) {
            for(let j = iR; j < R.length; j++){
                A[i + j - iR] = R[j];
            }
            break;
        }
        if (iR >= R.length) {
            for(let j = iL; j < L.length; j++){
                A[i + j - iL] = L[j];
            }
            break;
        }
        if (R[iR] > L[iL]){
            A[i] = L[iL]
            iL ++;
        } else {
            A[i] = R[iR];
            iR ++;
        }
    }
}

let A = [94, 4, 23, 12, 3, 43, 651, 12, 6, 9, 1];
sort(A);
console.log(A);
